# Summary

In this notebook, we will perform Exploratory Data Analysis (EDA) in Python. Steps:
1. Preliminary
2. Data cleaning / Feature Engineering / Data Manipulation
3. Reading and Extracting Data

# Import Libraries

In [22]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

# Load Data

In [23]:
# adjust the path accordingly
df = pd.read_csv('admission_data.csv')
data = df.copy()

#  Preliminary

In [24]:
# Showing top 5 rows of the data
data.head()

gre_score  toefl_score  univ_ranking  motiv_letter_strength  \
0      337.0        118.0           4.0                    4.5   
1      324.0        107.0           4.0                    4.0   
2      316.0        104.0           3.0                    3.0   
3      322.0        110.0           3.0                    3.5   
4      314.0        103.0           2.0                    2.0   

   recommendation_strength   gpa research_exp  admit_status  
0                      4.5  9.65          yes             1  
1                      4.5  8.87          yes             1  
2                      3.5  8.00          yes             1  
3                      2.5  8.67          yes             1  
4                      3.0  8.21           no             0

In [25]:
data.tail()

gre_score  toefl_score  univ_ranking  motiv_letter_strength  \
500      313.0        102.0           3.0                    3.5   
501      328.0        108.0           4.0                    4.5   
502      321.0        109.0           4.0                    4.0   
503      325.0        106.0           3.0                    3.5   
504      314.0        106.0           3.0                    3.0   

     recommendation_strength   gpa research_exp  admit_status  
500                      4.0  8.90          yes             1  
501                      4.0  9.18          yes             1  
502                      4.0  8.68          yes             0  
503                      4.0  8.40          yes             0  
504                      5.0  8.90           no             1

In [26]:
# sample 5 rows of the data
# the results change everytime the code is executed (random sample)
data.sample(5)

gre_score  toefl_score  univ_ranking  motiv_letter_strength  \
67       316.0        107.0           2.0                    3.5   
37       300.0        105.0           NaN                    1.0   
239      299.0        100.0           1.0                    1.5   
407      298.0        100.0           3.0                    2.5   
437      317.0        106.0           NaN                    1.5   

     recommendation_strength   gpa research_exp  admit_status  
67                       3.5  8.64          yes             0  
37                       2.0  7.80           no             0  
239                      2.0  7.89           no             0  
407                      4.0  7.95          yes             0  
437                      3.5  7.65          yes             0

Observations:
1. All columns, except `research_exp` column, are numeric
2. `research_exp` column seems to contain two distinct values (yes OR no), but will confirm later
3. `admit_status` is apparently also binary (0,1)
4. No obvious defect on the data (column name vs its entries), all looks good

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 505 entries, 0 to 504
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   gre_score                496 non-null    float64
 1   toefl_score              497 non-null    float64
 2   univ_ranking             498 non-null    float64
 3   motiv_letter_strength    503 non-null    float64
 4   recommendation_strength  499 non-null    float64
 5   gpa                      501 non-null    float64
 6   research_exp             505 non-null    object 
 7   admit_status             505 non-null    int64  
dtypes: float64(6), int64(1), object(1)
memory usage: 31.7+ KB


Observation:
1. Data contains 8 columns with 505 rows
2. Most of the columns have missing values (except `research_exp` and `admit_status`), they will be handled later
3. All dtypes seem OK (appropriate), given the corresponding column name

# Statistical Summary

In [28]:
data.columns

Index(['gre_score', 'toefl_score', 'univ_ranking', 'motiv_letter_strength',
       'recommendation_strength', 'gpa', 'research_exp', 'admit_status'],
      dtype='object')

In [29]:
# group column names based on type
# it will make our life easier onwards
categoricals = ['research_exp']

numericals = ['gre_score', 'toefl_score', 'univ_ranking',
              'motiv_letter_strength','recommendation_strength',
              'gpa','admit_status']

In [30]:
# Syntax numerical statistical summary
data[numericals].describe()

gre_score  toefl_score  univ_ranking  motiv_letter_strength  \
count  496.000000   497.000000    498.000000             503.000000   
mean   316.491935   107.173038      3.130522               3.382704   
std     11.218848     6.024667      1.135238               0.985532   
min    290.000000    92.000000      1.000000               1.000000   
25%    308.000000   103.000000      2.000000               2.500000   
50%    317.000000   107.000000      3.000000               3.500000   
75%    325.000000   112.000000      4.000000               4.000000   
max    340.000000   120.000000      5.000000               5.000000   

       recommendation_strength         gpa  admit_status  
count               499.000000  501.000000    505.000000  
mean                  3.484970    8.574571      0.574257  
std                   0.925582    0.600892      0.494945  
min                   1.000000    6.800000      0.000000  
25%                   3.000000    8.130000      0.000000  
50%                   3.500000    8.560000      1.000000  
75%                   4.000000    9.040000      1.000000  
max                   5.000000    9.920000      1.000000

Observation:
* Overall, the minimum and maximum values make sense for each column
* Mean ~ 50% (Median) in `gre_score`, `toefl_score`, and `gpa` column, indicating somewhat a symmetrical distribution
* `admit_status` column is boolean/binary column since the value is 0 or 1, no need to conclude its simmetricity. Only need to check balance level
* `univ_ranking`, `motiv_letter_strength`, and `recommendation_strength` are discrete values with not that many unique values, need to check values distribution later.


In [31]:
# Syntax describe method on categorical data
data[categoricals].describe()

research_exp
count           505
unique            2
top             yes
freq            280

Observations:
* `research_exp` have 2 unique value, 'yes', and 'no'
* student most have research_exp (280 rows), and the rest have no research experience.

In [32]:
# showing the precise value counts
# this code is especially useful if we have many categorical columns
for col in categoricals:
  print(f"Value counts of {col} column")
  print(data[col].value_counts(), '\n')

Value counts of research_exp column
yes    280
no     225
Name: research_exp, dtype: int64 



In [33]:
for col in data.columns:
  print(f"==== {col} ====")
  print(data[col].value_counts(), '\n')

==== gre_score ====
312.0    24
324.0    23
316.0    18
322.0    17
321.0    17
325.0    16
327.0    16
311.0    16
314.0    16
320.0    15
317.0    15
315.0    13
323.0    13
313.0    13
308.0    13
326.0    12
319.0    12
318.0    12
300.0    12
301.0    11
304.0    11
305.0    11
310.0    11
328.0    10
299.0    10
298.0    10
329.0    10
331.0     9
340.0     9
307.0     9
309.0     9
334.0     8
330.0     8
306.0     7
332.0     7
302.0     7
297.0     6
296.0     5
295.0     5
336.0     5
303.0     5
333.0     4
335.0     4
339.0     3
338.0     3
337.0     2
290.0     2
293.0     1
294.0     1
Name: gre_score, dtype: int64 

==== toefl_score ====
110.0    43
105.0    37
106.0    30
104.0    29
107.0    28
112.0    28
102.0    25
103.0    25
100.0    23
99.0     23
109.0    20
101.0    20
108.0    20
111.0    20
113.0    18
114.0    17
116.0    15
118.0    10
119.0    10
115.0    10
98.0      9
120.0     9
117.0     8
97.0      7
96.0      5
95.0      3
93.0      2
94.0      2
92

# Pembersihan Data

## 1. Duplicate Handling

In [34]:
len(data.drop_duplicates()) / len(data)
#jika output dari code di cell ini tidak bernilai 1 maka terdapat duplikat

0.9900990099009901

In [35]:
# Langkah 1: Ambil baris duplikat (termasuk yang asli)
duplicates = data[data.duplicated(keep=False)]

# Langkah 2: Hitung frekuensi kemunculan tiap baris duplikat
duplicate_counts = duplicates.groupby(list(data.columns)).size().reset_index(name='jumlah_duplikat')

# Langkah 3: Urutkan berdasarkan jumlah duplikat
sorted_duplicates = duplicate_counts.sort_values(by='jumlah_duplikat', ascending=False)

# Tampilkan hasil
print("Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:")
sorted_duplicates

Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:


gre_score  toefl_score  univ_ranking  motiv_letter_strength  \
0      313.0        102.0           3.0                    3.5   
1      314.0        106.0           3.0                    3.0   
2      321.0        109.0           4.0                    4.0   
3      325.0        106.0           3.0                    3.5   
4      328.0        108.0           4.0                    4.5   

   recommendation_strength   gpa research_exp  admit_status  jumlah_duplikat  
0                      4.0  8.90          yes             1                2  
1                      5.0  8.90           no             1                2  
2                      4.0  8.68          yes             0                2  
3                      4.0  8.40          yes             0                2  
4                      4.0  9.18          yes             1                2

Pada kode (df[df.duplicated(keep=False)]), Anda langsung mencetak baris yang memiliki duplikat dengan menggunakan metode duplicated dan memasukkan parameter keep=False. Ini mencetak semua baris yang memiliki duplikat tanpa mengurangi apa pun dari DataFrame.

In [36]:
#Handling Drop duplicate
data = data.drop_duplicates()

In [37]:
len(data.drop_duplicates()) / len(data)
#jika output dari code di cell ini tidak bernilai 1 maka terdapat duplikat

1.0

#### Telah dihandling drop duplikat

# 2. Missing value handling
### Identifying Missing value

In [38]:
data.isna().sum()

gre_score                  9
toefl_score                8
univ_ranking               7
motiv_letter_strength      2
recommendation_strength    6
gpa                        4
research_exp               0
admit_status               0
dtype: int64

In [39]:
data.isnull().sum()

gre_score                  9
toefl_score                8
univ_ranking               7
motiv_letter_strength      2
recommendation_strength    6
gpa                        4
research_exp               0
admit_status               0
dtype: int64

In [40]:
for column in data.columns:
    print(f"============= {column} =================")
    display(data[column].value_counts())
    print()

============= gre_score =================


312.0    24
324.0    23
316.0    18
322.0    17
321.0    16
327.0    16
311.0    16
314.0    15
320.0    15
325.0    15
317.0    15
308.0    13
323.0    13
315.0    13
319.0    12
318.0    12
313.0    12
326.0    12
300.0    12
304.0    11
305.0    11
301.0    11
310.0    11
299.0    10
329.0    10
298.0    10
331.0     9
340.0     9
307.0     9
328.0     9
309.0     9
334.0     8
330.0     8
306.0     7
332.0     7
302.0     7
297.0     6
296.0     5
295.0     5
336.0     5
303.0     5
333.0     4
335.0     4
339.0     3
338.0     3
337.0     2
290.0     2
293.0     1
294.0     1
Name: gre_score, dtype: int64


============= toefl_score =================


110.0    43
105.0    37
104.0    29
107.0    28
106.0    28
112.0    28
103.0    25
102.0    24
100.0    23
99.0     23
101.0    20
111.0    20
109.0    19
108.0    19
113.0    18
114.0    17
116.0    15
118.0    10
119.0    10
115.0    10
98.0      9
120.0     9
117.0     8
97.0      7
96.0      5
95.0      3
93.0      2
94.0      2
92.0      1
Name: toefl_score, dtype: int64


============= univ_ranking =================


3.0    161
2.0    123
4.0    104
5.0     73
1.0     32
Name: univ_ranking, dtype: int64


============= motiv_letter_strength =================


4.0    89
3.5    88
3.0    80
4.5    63
2.5    63
2.0    43
5.0    42
1.5    24
1.0     6
Name: motiv_letter_strength, dtype: int64


============= recommendation_strength =================


3.0    98
4.0    92
3.5    85
4.5    62
2.5    50
5.0    49
2.0    46
1.5    11
1.0     1
Name: recommendation_strength, dtype: int64


============= gpa =================


8.00    9
8.76    9
8.45    7
8.56    7
8.54    7
8.50    6
8.34    6
8.64    6
9.02    6
8.66    6
9.04    6
7.88    6
9.11    6
8.65    6
8.12    6
8.13    5
9.23    5
7.86    5
8.02    5
7.64    5
8.78    5
8.80    5
7.65    5
7.89    5
8.10    5
8.20    5
8.26    5
9.00    5
9.10    5
8.67    5
8.27    5
8.22    5
8.97    4
8.48    4
8.46    4
9.36    4
8.77    4
8.03    4
8.96    4
9.08    4
8.33    4
9.16    4
9.45    4
8.40    4
9.22    4
9.13    4
8.30    4
8.79    4
8.43    4
9.14    4
7.90    4
8.73    3
8.16    3
8.36    3
8.57    3
8.87    3
7.46    3
8.62    3
8.01    3
7.68    3
8.18    3
8.37    3
9.44    3
8.04    3
8.53    3
8.74    3
9.66    3
9.06    3
8.75    3
8.69    3
8.44    3
8.24    3
9.01    3
9.12    3
8.68    3
7.95    3
9.80    3
8.42    3
7.66    3
7.80    3
8.90    3
9.60    3
8.21    3
8.70    2
8.17    2
7.98    2
7.34    2
8.32    2
9.25    2
8.94    2
7.70    2
9.87    2
9.34    2
9.40    2
8.15    2
9.26    2
7.50    2
8.09    2
8.07    2
8.60    2



============= research_exp =================


yes    276
no     224
Name: research_exp, dtype: int64


============= admit_status =================


1    287
0    213
Name: admit_status, dtype: int64

- The GRE has a score range of 260–340 for the total score
- toefl score 0-120

In [41]:
# percentage version
total_rows = len(data)

# Menghitung dan menampilkan persentase missing values di setiap kolom satu per satu
for column in data.columns:
    missing_count = data[column].isna().sum()
    missing_percentage = (missing_count / total_rows) * 100
    print(f"Column '{column}' Has {missing_count} missing values ({missing_percentage:.2f}%)") # .2f means 2 decimal

Column 'gre_score' Has 9 missing values (1.80%)
Column 'toefl_score' Has 8 missing values (1.60%)
Column 'univ_ranking' Has 7 missing values (1.40%)
Column 'motiv_letter_strength' Has 2 missing values (0.40%)
Column 'recommendation_strength' Has 6 missing values (1.20%)
Column 'gpa' Has 4 missing values (0.80%)
Column 'research_exp' Has 0 missing values (0.00%)
Column 'admit_status' Has 0 missing values (0.00%)


The percentage of missing values below 20% so we handle numerically with median, categorical with mode. But the categorical data type does not have missing values, namely research_exp.

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   gre_score                491 non-null    float64
 1   toefl_score              492 non-null    float64
 2   univ_ranking             493 non-null    float64
 3   motiv_letter_strength    498 non-null    float64
 4   recommendation_strength  494 non-null    float64
 5   gpa                      496 non-null    float64
 6   research_exp             500 non-null    object 
 7   admit_status             500 non-null    int64  
dtypes: float64(6), int64(1), object(1)
memory usage: 35.2+ KB


In [43]:
# Handling missing value for EDA, without splitting
for column in data.columns:
    if data[column].dtype == 'object':
        # Jika kolom bertipe object, isi dengan mode
        data[column].fillna(data[column].mode()[0], inplace=True)
    else: # karena tipe data hanya object dan numerik saja tidak ada yg lain, maka pakai else berikut
        # Jika kolom bertipe numerik, isi dengan median
        data[column].fillna(data[column].median(), inplace=True)

In [44]:
data.isna().sum()

gre_score                  0
toefl_score                0
univ_ranking               0
motiv_letter_strength      0
recommendation_strength    0
gpa                        0
research_exp               0
admit_status               0
dtype: int64

In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   gre_score                500 non-null    float64
 1   toefl_score              500 non-null    float64
 2   univ_ranking             500 non-null    float64
 3   motiv_letter_strength    500 non-null    float64
 4   recommendation_strength  500 non-null    float64
 5   gpa                      500 non-null    float64
 6   research_exp             500 non-null    object 
 7   admit_status             500 non-null    int64  
dtypes: float64(6), int64(1), object(1)
memory usage: 35.2+ KB


Success removing NULL values!

## Thank you :)